// @progress/kendo-react-gauges/dist/npm/theming/theme-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var kendo_charts_1 = require("@progress/kendo-charts");
var template = "\n    <div class=\"k-var--gauge-pointer\"></div>\n    <div class=\"k-var--gauge-track\"></div>\n    <div class=\"k-var--normal-text-color\"></div>\n";
/**
 * @hidden
 */
var ThemeService = /** @class */ (function () {
    function ThemeService(store) {
        this.store = store;
    }
    ThemeService.prototype.setStyle = function (field, value) {
        this.store.dispatch({ type: 'set', payload: { field: field, value: value } });
    };
    ThemeService.prototype.setColors = function () {
        this.mapColor('pointer.color', 'gauge-pointer');
        this.mapColor('scale.rangePlaceholderColor', 'gauge-track');
        this.mapColor('scale.labels.color', 'normal-text-color');
        this.mapColor('scale.minorTicks.color', 'normal-text-color');
        this.mapColor('scale.majorTicks.color', 'normal-text-color');
        this.mapColor('scale.line.color', 'normal-text-color');
    };
    ThemeService.prototype.mapColor = function (key, varName) {
        this.setStyle(key, this.queryStyle(varName).backgroundColor);
    };
    ThemeService.prototype.queryStyle = function (varName) {
        var element = this.element;
        return window.getComputedStyle(element.querySelector(".k-var--" + varName));
    };
    return ThemeService;
}());
var loadTheme = function (store, done) {
    var service = new ThemeService(store);
    if (typeof document === 'undefined') {
        store.dispatch({ type: 'push', payload: kendo_charts_1.chartBaseTheme() });
        return;
    }
    var container = service.element = document.createElement('div');
    container.style.display = 'none';
    container.innerHTML = template;
    document.body.appendChild(container);
    try {
        store.dispatch({ type: 'push', payload: kendo_charts_1.chartBaseTheme() });
        service.setColors();
    }
    finally {
        document.body.removeChild(service.element);
        delete service.element;
        done();
    }
};
exports.loadTheme = loadTheme;
